### DISCUSSION TITLE
Does prompt-toolkit have clipboard buffers?

### DISCUSSION THREAD

**Discussion by Spitfire1900** [prompt-toolkit/python-prompt-toolkit#1939] [Q&A]
I use Xonsh and they directed me here as it uses prompt-toolkit underneath.

I frequently with copy a string from an IDE or web browser to paste it into a Xonsh session in VI mode. I'll use normal mode's `c+w` or `d+t+<char>` prior to pasting. This overwrites the clipboard buffer.

is there a good approach to keep buffers separate so deleting/yanking in prompt-toolkit does not replace the system clipboard?

**Comment by TayfurYldz** [comment]
prompt-toolkit does have its own clipboard abstraction, and its default clipboard is already separate from the OS clipboard.

`PromptSession` creates an `InMemoryClipboard` by default. That implementation keeps a kill ring in memory (up to 60 entries by default), so vi-mode deletes/yanks operate on prompt-toolkit's clipboard rather than the Windows/macOS/Linux clipboard.

The system-synchronized behavior only happens if the application is configured with `PyperclipClipboard`, whose `set_data()` calls `pyperclip.copy()`.

So for the behavior you describe, make sure the Xonsh integration is using an `InMemoryClipboard` rather than `PyperclipClipboard`. At the prompt-toolkit API level that is simply:

```python
from prompt_toolkit.clipboard import InMemoryClipboard
from prompt_toolkit import PromptSession

session = PromptSession(clipboard=InMemoryClipboard())
```

There is also a useful distinction here: terminal bracketed-paste / OS paste can still insert text into the prompt even though prompt-toolkit's vi kill/yank register is independent. Separating the clipboard therefore doesn't prevent normal Ctrl/terminal paste; it prevents prompt-toolkit's own delete/yank operations from overwriting the system clipboard.

For Xonsh specifically, the remaining question is which clipboard implementation its prompt-toolkit shell instance supplies; that configuration belongs on the Xonsh side, but prompt-toolkit itself supports the separation directly.
